src/files/file-preview/FilePreview.js

```diff
@@ -18,20 +20,43 @@ const Preview = (props) => {
 
   const type = typeFromExt(name)
 
-  return <div className={ classNames(type !== 'pdf' && 'dib') } ref={drag}>
+  // Hack: Allows for text selection if it's a text file (bypass useDrag)
+  const dummyRef = useRef()
+
+  return <div className={ classNames(type !== 'pdf' && type !== 'text' && type !== 'json' && 'dib') } ref={type === 'text' ? dummyRef : drag}>
     <PreviewItem {...props} type={type} />
   </div>
 }
 
-const PreviewItem = ({ t, name, cid, size, type, availableGatewayUrl: gatewayUrl, read }) => {
+const PreviewItem = ({ t, name, cid, size, type, availableGatewayUrl: gatewayUrl, read, onDownload }) => {
   const [content, setContent] = useState(null)
+  const [hasMoreContent, setHasMoreContent] = useState(false)
+  const [buffer, setBuffer] = useState(null)
 
-  const loadContent = async () => {
-    const buf = await read()
-    setContent(buf.toString('utf-8'))
-  }
+  const loadContent = useCallback(async () => {
+    const readBuffer = buffer || await read()
+    if (!buffer) {
+      setBuffer(readBuffer)
+    }
+
+    const { value, done } = await readBuffer.next()
+    const previousContent = content || ''
+
+    const currentContent = previousContent + fromUint8ArrayToString(value)
+
+    setContent(currentContent)
+
+    const hasMore = !done && new TextEncoder().encode(currentContent).length < size
 
-  const src = `${gatewayUrl}/ipfs/${cid}`
+    setHasMoreContent(hasMore)
+  }, [buffer, content, read, size])
+
+  useEffect(() => {
+    loadContent()
+  }, // eslint-disable-next-line react-hooks/exhaustive-deps
+  [])
+
+  const src = `${gatewayUrl}/ipfs/${cid}?filename=${encodeURIComponent(name)}`
```

Comment:
FYI I've added `filename` here, so its preserved when "Download" is clicked or when someone tries to download via browser's context menu.